stdlib/public/runtime/DynamicCast.cpp

```diff
@@ -591,6 +591,16 @@ tryCastToForeignClass(
   return DynamicCastResult::Failure;
 }
 
+static DynamicCastResult tryCastToForeignReferenceType(
+    OpaqueValue *destLocation, const Metadata *destType, OpaqueValue *srcValue,
+    const Metadata *srcType, const Metadata *&destFailureType,
+    const Metadata *&srcFailureType, bool takeOnSuccess, bool mayDeferChecks) {
+  assert(srcType != destType);
+  assert(destType->getKind() == MetadataKind::ForeignReferenceType);
+
+  return DynamicCastResult::Failure;
```

Comment:
Despite this returning `Failure` unconditionally, the cast eventually succeeds in this code path:
https://github.com/swiftlang/swift/blob/84ae5fbe8c2f7e3724bf07551167656f3da3b1dc/stdlib/public/runtime/DynamicCast.cpp#L2373-L2381

Reply:
Nice!   Yeah, I think I remember there are other cases where an always-failing conversion is the right first step to unblock later progress.  It might help people to think of the dynamic caster as a search process that evaluates a variety of paths to find a feasible conversion.

I _suppose_ we could consider having `selectCasterForDest` fall back to returning a `tryCastUnknown` that unconditionally fails, instead of falling through to an unreachable assertion.  That would allow many of the general engines (optional unwrapping, existential unwrapping) to work even for otherwise-unrecognized types.